vllm/worker/model_runner.py

```diff
@@ -168,7 +167,91 @@ def from_broadcasted_tensor_dict(
 
 
 class ModelInputForGPUBuilder(ModelRunnerInputBuilderBase[ModelInputForGPU]):
-    """TBA"""
+    """Build ModelInputForGPU from SequenceGroupMetadata."""
+
+    # Compute function names for each sequence in
+    # a sequence group. The order of the functions matters
+    PER_SEQ_COMPUTE_FN_NAMES = (
+        "_compute_lens",
+        "_compute_for_prefix_cache_hit",
+        "_compute_for_sliding_window",
+        "_compute_lora_input",
+    )
+    # Compute function names for each sequence group.
+    # The order of the functions matters
+    PER_SEQ_GROUP_COMPUTE_FN_NAMES = (
+        "_compute_prompt_adapter_input",
+        "_compute_multi_modal_input",
+    )
+
+    @dataclasses.dataclass
+    class InterDataForSeqGroup:
+        """Intermediate data for the current sequence group."""
+        # From sequence group metadata.
+        request_id: str
+        seq_ids: List[int]
+        n_seqs: int
+        is_prompt: bool
+        block_tables: Optional[Dict[int, List[int]]]
+        computed_block_nums: List[int]
+
+        # Input tokens and positions.
+        input_tokens: List[List[int]]
+        input_positions: List[List[int]]
+
+        # The sequence length (may be capped to the sliding window).
+        seq_lens: List[int]
+        # The original sequence length (before applying sliding window).
+        # This is used to compute slot mapping.
+        orig_seq_lens: List[int]
+        # The query length.
+        query_lens: List[int]
+        # The number of tokens that are already computed.
+        context_lens: List[int]
+        # The current sliding window block.
+        curr_sliding_window_blocks: List[int]
+
+        # LoRA inputs.
+        lora_index_mapping: List[List[int]]
+        lora_prompt_mapping: List[List[int]]
+        lora_requests: Set[LoRARequest]
+
+        # Prompt adapter inputs.
+        prompt_adapter_index_mapping: List[int]
+        prompt_adapter_prompt_mapping: List[int]
+        prompt_adapter_request: Optional[PromptAdapterRequest] = None
+
+        # Multi-modal inputs.
+        multi_modal_inputs: Optional[MultiModalInputs] = None
+
+        # Whether the prefix cache is hit (prefill only).
+        prefix_cache_hit: bool = False
+
+        def __init__(self, request_id: str, seq_ids: List[int],
```

Comment:
why do we need to override `__init__`? that's an anti-pattern for dataclasses. we should use `__post_init__` instead or not use a dataclass in the first place.

Reply:
I did that because I wish to initialize them with `n_seqs`, but you're right I can still do that with `__post_init__`